File: tools/probe-no-reload.mjs
// Acceptance probe for the no-reload work.
//   A  an unrelated interaction must not destroy the viewer's iframes
//   B  navigation done INSIDE a live tile must survive an unrelated interaction  <-- the user's complaint
//   C  browser/morph capability facts
//   D  morph hazards: duplicate ids, <details> open state, typed text, web components
// Run against a live studio:
//   dart run appboxd/bin/appbox.dart design serve designs/appbox-studio --port 4319 --json --project portalo
//   node tools/probe-no-reload.mjs
// Override with APPBOX_BASE / APPBOX_CHROME. Exits non-zero if a check fails.
import { fileURLToPath } from 'node:url';
import path from 'node:path';
const REPO = path.resolve(path.dirname(fileURLToPath(import.meta.url)), '..');
const { chromium } = await import(
  path.join(REPO, 'skills/appbox-designer/runtime/node_modules/playwright-core/index.mjs'));

const CHROME = process.env.APPBOX_CHROME || '/Applications/Google Chrome.app/Contents/MacOS/Google Chrome';
const BASE = process.env.APPBOX_BASE || 'http://localhost:4319';
const settle = (p) => p.waitForTimeout(900);
const ok = (b) => (b ? 'PASS' : 'FAIL');

const mark = `(() => { const f=[...document.querySelectorAll('iframe')]; f.forEach((x,i)=>x.__probe='p'+i); return f.length; })()`;
const survivors = `(() => { const f=[...document.querySelectorAll('iframe')]; return {total:f.length, marked:f.filter(x=>x.__probe!==undefined).length}; })()`;
const dupIds = `(() => {
  const seen={}, dup=[];
  for (const el of document.querySelectorAll('[id]')) { if (seen[el.id]) dup.push(el.id); seen[el.id]=1; }
  return [...new Set(dup)];
})()`;

let browser, fails = 0;
const check = (name, pass, extra = '') => { if (!pass) fails++; console.log(`  [${ok(pass)}] ${name}${extra ? ' — ' + extra : ''}`); };

try {
  browser = await chromium.launch({ executablePath: CHROME, headless: true });
  const page = await browser.newPage({ viewport: { width: 1600, height: 1000 } });
  const errs = [];
  page.on('pageerror', (e) => errs.push(e.message));
  let navs = [];
  page.on('framenavigated', (f) => { if (f !== page.mainFrame()) navs.push(f.url().replace(BASE, '')); });

  await page.goto(BASE + '/design', { waitUntil: 'networkidle' });
  await page.waitForSelector('iframe.dv-tile-frame', { timeout: 15000 });
  await settle(page);

  console.log('=== C. capability ===');
  const morphReady = await page.evaluate(`!!(window.htmx && htmx.config && document.querySelector('[hx-ext*="morph"]')) && typeof Idiomorph !== 'undefined'`);
  console.log('  htmx            :', await page.evaluate('window.htmx && htmx.version'));
  console.log('  Idiomorph loaded:', await page.evaluate(`typeof Idiomorph`));
  console.log('  Element.moveBefore:', await page.evaluate(`'moveBefore' in Element.prototype`));
  check('morph extension wired', morphReady);

  console.log('\n=== D1. duplicate ids (morph correctness precondition) ===');
  const dups = await page.evaluate(dupIds);
  check('no duplicate ids in /design', dups.length === 0, dups.length ? dups.slice(0, 6).join(', ') : '');

  console.log('\n=== A. does an unrelated interaction destroy the iframes? ===');
  const before = await page.evaluate(mark);
  navs = [];
  await page.$eval('.dv-tile .dv-tile-tools a[hx-get*="context"]', (el) => el.click());
  await settle(page);
  const after = await page.evaluate(survivors);
  console.log(`  iframes ${before} -> surviving DOM nodes ${after.marked}, iframe navigations ${navs.length}`);
  check('iframes survive as the same DOM nodes', after.marked >= before - 1, `${after.marked}/${before}`);
  check('no iframe re-navigation', navs.length === 0, `${navs.length} navigations`);

  console.log('\n=== D2. <details> open state + typed text survive a swap ===');
  await page.evaluate(`(() => { const d=document.querySelector('details.dv-tool-menu'); if(d) d.open=true; })()`);
  const hadDetails = await page.evaluate(`!!document.querySelector('details.dv-tool-menu')`);
  await page.evaluate(`(() => { const t=document.querySelector('.composer-card textarea, textarea[name="text"]'); if(t){ t.value='DRAFT-KEEP-ME'; } })()`);
  const hadTextarea = await page.evaluate(`!!document.querySelector('textarea[name="text"]')`);
  await page.$eval('.dv-tile[data-id="portalo.cart"] .dv-tile-tools a[hx-get*="context"]', (el) => el.click());
  await settle(page);
  // DEFERRED, not fixed by morph, and NOT a regression — verified by stashing
  // the templates and re-running: at baseline these nodes were DESTROYED
  // (sameNode:false) and lost the same state. Morph now keeps the node but
  // still applies the server's content, and the server echoes neither `open`
  // nor the draft text. The real fix is Lever 2 (a pin toggle should not
  // re-render the composer at all), not a bigger hammer here. Reported, and
  // deliberately excluded from this slice's pass/fail.
  const detailsOpen = hadDetails && await page.evaluate(`(()=>{const d=document.querySelector('details.dv-tool-menu'); return !!(d&&d.open);})()`);
  const textKept = hadTextarea && await page.evaluate(`(()=>{const t=document.querySelector('textarea[name="text"]'); return !!(t&&t.value==='DRAFT-KEEP-ME');})()`);
  // Fixed: the textarea is hx-preserve'd (dropped only on the render after a
  // send, so the sent text cannot linger and be sent twice). tools/… has no
  // send coverage; that pairing is asserted separately.
  if (hadTextarea) check('typed composer draft preserved', textKept);
  // Still open: the server echoes <details> without `open`, so morph closes it.
  // Not a regression — at baseline the node was destroyed outright.
  console.log(`  [KNOWN-OPEN] <details> stayed open      : ${detailsOpen} (baseline: false — node destroyed)`);
  console.log('               server does not echo `open`; morph applies server state faithfully.');

  console.log('\n=== B. navigation INSIDE a live tile survives (THE complaint) ===');
  await page.$eval('.dv-tile[data-id="portalo.home"] a[hx-get*="live="]', (el) => el.click());
  await settle(page);
  const liveSel = '.dv-tile[data-id="portalo.home"] iframe';
  let frame = await (await page.$(liveSel)).contentFrame();
  console.log('  starts at :', frame.url().replace(BASE, ''));
  const link = await frame.$('a[href*="portalo.category"]');
  if (!link) { check('found in-frame link to navigate', false); }
  else {
    await link.click();
    await settle(page);
    frame = await (await page.$(liveSel)).contentFrame();
    console.log('  after nav :', frame.url().replace(BASE, ''));
    await page.$eval('.dv-tile[data-id="portalo.cart"] .dv-tile-tools a[hx-get*="context"]', (el) => el.click());
    await settle(page);
    frame = await (await page.$(liveSel)).contentFrame();
    const ended = frame.url().replace(BASE, '');
    console.log('  after unrelated pin :', ended);
    check('live tile kept the user\'s navigation', ended.includes('portalo.category'), ended);
  }

  console.log('\n=== D3. web components still upgraded after a morph ===');
  const wc = await page.evaluate(`(() => {
    const names = ['model-viewer','dotlottie-wc'];
    const out = {};
    for (const n of names) {
      const el = document.querySelector(n);
      out[n] = el ? (customElements.get(n) ? 'defined' : 'present-undefined') : 'absent-in-shell';
    }
    return out;
  })()`);
  console.log('  ', JSON.stringify(wc), '(shell-level; portalo web components live inside iframes)');

  console.log('\n=== page errors ===');
  console.log(errs.length ? errs.slice(0, 5).map(e => '  ! ' + e).join('\n') : '  none');
  check('no uncaught page errors', errs.length === 0);
} catch (e) {
  console.log('PROBE ERROR:', e.message); fails++;
} finally {
  if (browser) await browser.close();
  console.log(`\n==== ${fails === 0 ? 'ALL CHECKS PASSED' : fails + ' CHECK(S) FAILED'} ====`);
  process.exit(fails === 0 ? 0 : 1);
}

File: arxa/lib/probes/studio/probe_no_reload.dart
// probe-no-reload — acceptance probe for the no-reload work.
//   A  an unrelated interaction must not destroy the viewer's iframes
//   B  navigation done INSIDE a live tile must survive an unrelated
//      interaction  <-- the user's complaint
//   C  browser/morph capability facts
//   D  morph hazards: duplicate ids, <details> open state, typed text, web
//      components
//
// Dart port of `archives/tooling-pre-dart/tools/studio-probes/probe-no-reload.mjs` (10 checks in the pass case, one
// page for the whole run — state has to carry across sections).

import 'dart:convert';

import 'package:arxa/probes/probe_base.dart';

const String _mark = r'''(() => { const f=[...document.querySelectorAll('iframe')]; f.forEach((x,i)=>x.__probe='p'+i); return f.length; })()''';
const String _survivors = r'''(() => { const f=[...document.querySelectorAll('iframe')]; return {total:f.length, marked:f.filter(x=>x.__probe!==undefined).length}; })()''';
const String _dupIds = r'''(() => {
  const seen={}, dup=[];
  for (const el of document.querySelectorAll('[id]')) { if (seen[el.id]) dup.push(el.id); seen[el.id]=1; }
  return [...new Set(dup)];
})()''';

const String _liveSel = '.dv-tile[data-id="portalo.home"] iframe';
const String _inFrameLink = 'a[href*="portalo.category"]';

const Probe noReloadProbe = Probe(
  name: 'no-reload',
  summary:
      'an unrelated pin must not destroy iframes or lose in-frame navigation',
  // Sections A, D2 and B click controls that POST (`hx-get*="context"`, an
  // in-tile navigation). The .mjs original never guards this — see
  // docs/probes-capability-map.md, "Registry — adding a probe (wave C)": the
  // port declares the guard anyway, which is a strengthening, not drift.
  mutates: true,
  body: _run,
);

Future<void> _run(ProbeContext ctx) async {
  final page = await ctx.newPage();
  await ctx.goto(page, '/design');
  await page.waitForSelector('iframe.dv-tile-frame',
      timeout: const Duration(seconds: 15));
  await waitQuiet(page, report: ctx.report);

  ctx.report.section('C. capability');
  // htmx4 ships morph in core (config.morphSkip/morphScanLimit, swap token
  // `outerMorph`) — there is no idiomorph extension and no `hx-ext` wiring
  // anymore. The old assertion encoded the v2 stack; asserting it against
  // htmx4 fails on every correctly-served page.
  final morphReady = await page.evaluate(
      "!!(window.htmx && htmx.config && 'morphSkip' in htmx.config && 'morphScanLimit' in htmx.config)");
  final htmxVersion = await page.evaluate('window.htmx && htmx.version');
  ctx.report.out.writeln('  htmx            : ${htmxVersion ?? 'undefined'}');
  final idiomorphType = await page.evaluate('typeof Idiomorph');
  ctx.report.out.writeln('  Idiomorph loaded: $idiomorphType (htmx4 core morph needs none)');
  final moveBefore = await page.evaluate("'moveBefore' in Element.prototype");
  ctx.report.out.writeln('  Element.moveBefore: $moveBefore');
  ctx.report.check('htmx4 core morph available', morphReady == true);

  ctx.report.section('D1. duplicate ids (morph correctness precondition)');
  final dups =
      ((await page.evaluate(_dupIds)) as List?)?.cast<String>() ?? const [];
  ctx.report.check('no duplicate ids in /design', dups.isEmpty,
      dups.isNotEmpty ? dups.take(6).join(', ') : '');

  // Started here, not before goto: the only checks that read it (E) exercise
  // the `/context/` GET the pin control fires, which first happens in A —
  // recording from the initial page load is unnecessary traffic to fetch and
  // buffer.
  final bodies = await page.recordNetworkBodies();

  ctx.report.section('A. does an unrelated interaction destroy the iframes?');
  final beforeCount = ((await page.evaluate(_mark)) as num).toInt();
  var navCount = 0;
  final mainFrameId = (await page.frames()).first.id;
  final navSub = page.on('Page.frameNavigated').listen((e) {
    final id = '${(e.params['frame'] as Map?)?['id']}';
    if (id != mainFrameId) navCount++;
  });
  // Synthetic: the tile tool rail is hover-revealed, so a real pointer cannot
  // reach it in a headless run.
  await page.clickSelector('.dv-tile .dv-tile-tools a[hx-get*="context"]',
      synthetic: true);
  await waitQuiet(page, report: ctx.report);
  await navSub.cancel();
  final afterRaw = (await page.evaluate(_survivors)) as Map;
  final afterMarked = (afterRaw['marked'] as num).toInt();
  ctx.report.out.writeln(
      '  iframes $beforeCount -> surviving DOM nodes $afterMarked, iframe navigations $navCount');
  ctx.report.check('iframes survive as the same DOM nodes',
      afterMarked >= beforeCount - 1, '$afterMarked/$beforeCount');
  ctx.report
      .check('no iframe re-navigation', navCount == 0, '$navCount navigations');

  ctx.report.section('D2. node identity + typed text survive a swap');
  final hadDetails = (await page.evaluate(r'''(() => { const d=document.querySelector('details.dv-tool-menu'); if(!d) return false; d.__probeNode=1; return true; })()''')) ==
      true;
  await page.evaluate(r'''(() => { const t=document.querySelector('.composer-card textarea, textarea[name="text"]'); if(t){ t.value='DRAFT-KEEP-ME'; } })()''');
  final hadTextarea =
      (await page.evaluate('''!!document.querySelector('textarea[name="text"]')''')) ==
          true;
  await page.clickSelector(
      '.dv-tile[data-id="portalo.cart"] .dv-tile-tools a[hx-get*="context"]',
      synthetic: true);
  await waitQuiet(page, report: ctx.report);
  final detailsSameNode = hadDetails &&
      (await page.evaluate(r'''(()=>{const d=document.querySelector('details.dv-tool-menu'); return !!(d&&d.__probeNode===1);})()''')) ==
          true;
  final textKept = hadTextarea &&
      (await page.evaluate(r'''(()=>{const t=document.querySelector('textarea[name="text"]'); return !!(t&&t.value==='DRAFT-KEEP-ME');})()''')) ==
          true;
  if (hadDetails) {
    ctx.report.check('<details> menu survives as the same node', detailsSameNode);
  }
  if (hadTextarea) {
    ctx.report.check('typed composer draft preserved', textKept);
  }

  ctx.report.section("B. navigation INSIDE a live tile survives (THE complaint)");
  await page.evaluate(
      "window.htmx.ajax('GET', '/design/viewer?mode=flows', { target: '#design-viewer', swap: 'morph:outerHTML' })");
  await waitQuiet(page, report: ctx.report);
  await page.clickSelector('.dv-tile[data-id="portalo.home"] a[hx-get*="step="]',
      synthetic: true);
  await waitQuiet(page, report: ctx.report);

  // Unguarded: the .mjs original has no null check resolving this iframe
  // either (`(await page.$(liveSel)).contentFrame()` throws on a miss), so a
  // miss here aborts the run the same way — same failure shape, same count.
  var frame = await page.frameForSelector(_liveSel);
  if (frame == null) {
    throw StateError('no live iframe at $_liveSel — the home tile never armed');
  }
  final startUrl = ((await page.evaluateInFrame(frame, 'location.href')) as String)
      .replaceFirst(ctx.base, '');
  ctx.report.out.writeln('  starts at : $startUrl');
  final hasLink = (await page.evaluateInFrame(frame,
          "document.querySelector(${jsonEncode(_inFrameLink)}) !== null")) ==
      true;
  if (!hasLink) {
    ctx.report.check('found in-frame link to navigate', false);
  } else {
    // Non-synthetic: the .mjs original clicks a real Playwright ElementHandle
    // here, which does hit-testing — the assertion below depends on it.
    await page.clickSelectorInFrame(frame, _inFrameLink);
    await waitQuiet(page, report: ctx.report);
    frame = await page.frameForSelector(_liveSel);
    if (frame == null) {
      throw StateError('no live iframe at $_liveSel after the in-frame navigation');
    }
    final afterNav = ((await page.evaluateInFrame(frame, 'location.href')) as String)
        .replaceFirst(ctx.base, '');
    ctx.report.out.writeln('  after nav : $afterNav');
    // Synthetic: same hover-revealed tile tool rail as section A.
    await page.clickSelector(
        '.dv-tile[data-id="portalo.cart"] .dv-tile-tools a[hx-get*="context"]',
        synthetic: true);
    await waitQuiet(page, report: ctx.report);
    frame = await page.frameForSelector(_liveSel);
    if (frame == null) {
      throw StateError('no live iframe at $_liveSel after the unrelated pin');
    }
    final ended = ((await page.evaluateInFrame(frame, 'location.href')) as String)
        .replaceFirst(ctx.base, '');
    ctx.report.out.writeln('  after unrelated pin : $ended');
    ctx.report.check("live tile kept the user's navigation",
        ended.contains('portalo.category'), ended);
  }

  ctx.report.section('D3. web components still upgraded after a morph');
  final wc = (await page.evaluate(r'''(() => {
  const names = ['model-viewer','dotlottie-wc'];
  const out = {};
  for (const n of names) {
    const el = document.querySelector(n);
    out[n] = el ? (customElements.get(n) ? 'defined' : 'present-undefined') : 'absent-in-shell';
  }
  return out;
})()''')) as Map;
  ctx.report.out.writeln(
      '   ${jsonEncode(wc)} (shell-level; portalo web components live inside iframes)');

  ctx.report.section('E. pin payload cost');
  // encodedLength (wire bytes), NOT the decoded body length the .mjs
  // original reads — a payload-size ceiling should read what actually
  // crossed the wire, not what gzip decompressed to. Deliberate
  // strengthening; recorded in the capability map.
  final ctxRecords = (await bodies.stop())
      .where((r) => r.url.contains('/context/') && r.encodedLength != null)
      .toList();
  final worstPin = ctxRecords.isEmpty
      ? 0
      : ctxRecords.map((r) => r.encodedLength!).reduce((a, b) => a > b ? a : b);
  ctx.report.out
      .writeln('  pin responses: ${ctxRecords.length}, largest $worstPin bytes');
  if (ctxRecords.isNotEmpty) {
    ctx.report.check('pin response under the 200 KB blow-up ceiling',
        worstPin < 200000, '$worstPin bytes');
  }

  ctx.report.section('page errors');
  final errs = page.pageErrors;
  ctx.report.out
      .writeln(errs.isEmpty ? '  none' : errs.take(5).map((e) => '  ! $e').join('\n'));
  ctx.report.check('no uncaught page errors', errs.isEmpty);

  await ctx.closePage(page);
}
